### PR TITLE
Make function rest parameters final

### LINKED ISSUE #41833: [Bug]: Function rest param is not set as final
### Description

$subject

According to the spec every parameter in the function should be final. But right now, the rest parameter can change. This means developers can modify them, which shouldn't happen according to the spec.

After the fix it should give a compile time error saying.
```
cannot assign a value to function argument 'restParam'(BCE2549)
```

### Steps to Reproduce

```ballerina
function name(string param1, string... rest) {
    rest = ["value"]; // no error
}
```

### Affected Version(s)

2201.8.2

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes #41833

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/finaltypes/test_implicitly_final_negative.bal [hunk 1]
```diff
@@ -44,3 +44,7 @@ listener test:MockListener ml = new (8080);
 public function testChangingListenerVariableAfterDefining() {
     ml = new test:MockListener(8081);
 }
+
+function restFinal(string p1, string... p2) {
+    p2 = ["a", "b"];
+}
```
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/finaltypes/test_implicitly_final_negative.bal:50]: Updated in [85eab29](https://github.com/ballerina-platform/ballerina-lang/pull/41844/commits/85eab29374b420659daef50aa0116c35d64971bc).

File: tests/jballerina-unit-test/src/test/resources/test-src/types/finaltypes/test_implicitly_final_negative.bal [hunk 2]
```diff
@@ -44,3 +44,7 @@ listener test:MockListener ml = new (8080);
 public function testChangingListenerVariableAfterDefining() {
     ml = new test:MockListener(8081);
 }
+
+function restFinal(string p1, string... p2) {
```
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/finaltypes/test_implicitly_final_negative.bal]: Updated in [85eab29](https://github.com/ballerina-platform/ballerina-lang/pull/41844/commits/85eab29374b420659daef50aa0116c35d64971bc).

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 3]
```diff
@@ -497,6 +497,10 @@ public void visit(BLangFunction funcNode, AnalyzerData data) {
         // TODO: Shouldn't this be done in symbol enter?
         //set function param flag to final
         funcNode.symbol.params.forEach(param -> param.flags |= Flags.FUNCTION_FINAL);
+        
+        if (funcNode.symbol.restParam != null) {
+            funcNode.symbol.restParam.flags |= Flags.FUNCTION_FINAL;
+        }
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:503]: This is not related to this change. What is the reason to introduce a new flag for this? Within the parameter symbol, there is already a flag to check whether it is a parameter. Can't we utilize that existing flag instead of introducing a new one? @MaryamZi

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:503]: Not sure which came first, but yeah, we may not need a separate flag. I wouldn't change this now though.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 4]
```diff
@@ -497,6 +497,10 @@ public void visit(BLangFunction funcNode, AnalyzerData data) {
         // TODO: Shouldn't this be done in symbol enter?
         //set function param flag to final
         funcNode.symbol.params.forEach(param -> param.flags |= Flags.FUNCTION_FINAL);
+        
+        if (funcNode.symbol.restParam != null) {
+            funcNode.symbol.restParam.flags |= Flags.FUNCTION_FINAL;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:502]: Extract `funcNode.symbol.restParam` out to a variable?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/finaltypes/test_implicitly_final_negative.bal [hunk 5]
```diff
@@ -44,3 +44,17 @@ listener test:MockListener ml = new (8080);
...
+function testRestParamFinal(string p1, string... p2) {
+    p2 = ["a", "b"];
+}
+
+function (int a, int... b) testModuleLevelRestParamFinal = function (int i, int... b) {
+        b = [];
+    };
+
+public function testLocalLevelRestParamFinal() {
+    function (int a, int... b) func = function (int i, int... b) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/finaltypes/test_implicitly_final_negative.bal:57]: One of these can use var.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/finaltypes/test_implicitly_final_negative.bal]: Do we have one with var? I meant `var f = ...`.

**Review by chiranSachintha — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/types/finaltypes/test_implicitly_final_negative.bal (hunk 5 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/finaltypes/test_implicitly_final_negative.bal]: Added in commit [#bf890dc](https://github.com/ballerina-platform/ballerina-lang/pull/41844/commits/bf890dc26035cc346247f1d2154c27869439a44f)

**Review by KavinduZoysa — APPROVED** [review]

**Review by MaryamZi — APPROVED** [review]
